### ISSUE TITLE
TypeError in flutil.manhattan_plot()

### ISSUE THREAD

**Issue by vb-iman** [fastlmm/FaST-LMM#41]
Hello! I have noticed an issue with using the flutil.manhattan_plot() function. Specifically, the issue seems to start with [this line](https://github.com/fastlmm/FaST-LMM/blob/master/fastlmm/util/util.py#L611) in def _compute_x_positions_chrom() in util.py where chromosome_starts is set as type "object", which leads to a downstream error with plt.xticks:

```
def _compute_x_positions_chrom(positions, offset=1e5):
    chromosomes = np.unique(positions[:, 0])
    chromosomes.sort()
    chromosome_starts = np.zeros((chromosomes.shape[0], 3), dtype="object")
    chr_start_next = 0
    for i, chromosome in enumerate(chromosomes):
        pos_chr = positions[positions[:, 0] == chromosome]
        chromosome_starts[i, 0] = chromosome  # the chromosome
        chromosome_starts[i, 1] = chr_start_next  # start of the chromosome
        chromosome_starts[i, 2] = (
            chr_start_next + pos_chr.max()
        )  # end of the chromosome
        chr_start_next = chromosome_starts[i, 2] + offset
    return chromosome_starts
```

I have been getting the following error:

TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

I was able to get around this by changing the dtype in line 611 to "int", but I wanted to raise the issue to you in case you had a better fix.

**Comment by CarlKCarlK** [comment]
Thanks! I'm working on FaST-LMM right now and will look into this.

From: vb-iman ***@***.***>
Sent: Tuesday, April 23, 2024 3:36 PM
To: fastlmm/FaST-LMM ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [fastlmm/FaST-LMM] TypeError in flutil.manhattan_plot() (Issue #41)
Importance: High


Hello! I have noticed an issue with using the flutil.manhattan_plot() function. Specifically, the issue seems to start with this line<https://github.com/fastlmm/FaST-LMM/blob/master/fastlmm/util/util.py#L611> in def _compute_x_positions_chrom() in util.py where chromosome_starts is set as type "object":

def _compute_x_positions_chrom(positions, offset=1e5):

    chromosomes = np.unique(positions[:, 0])

    chromosomes.sort()

    chromosome_starts = np.zeros((chromosomes.shape[0], 3), dtype="object")

    chr_start_next = 0

    for i, chromosome in enumerate(chromosomes):

        pos_chr = positions[positions[:, 0] == chromosome]

        chromosome_starts[i, 0] = chromosome  # the chromosome

        chromosome_starts[i, 1] = chr_start_next  # start of the chromosome

        chromosome_starts[i, 2] = (

            chr_start_next + pos_chr.max()

        )  # end of the chromosome

        chr_start_next = chromosome_starts[i, 2] + offset

    return chromosome_starts

I have been getting the following error:

TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

I was able to get around this by changing the dtype in line 611 to "int", but I wanted to raise the issue to you in case you had a better fix.

-

**Comment by CarlKCarlK** [comment]
Thanks for finding this bug with working with the newer version of numpy!

Here is the fix I'm looking at. If you want to try this, I'd love confirmation that it works as you expected. I've marked the two changes with `###`

```python
def _compute_x_positions_chrom(positions, offset=1e5):
    chromosomes = np.unique(positions[:, 0])
    chromosomes.sort()
    chromosome_starts = np.zeros((chromosomes.shape[0], 3), dtype="float") #### float
    chr_start_next = 0
    for i, chromosome in enumerate(chromosomes):
        pos_chr = positions[positions[:, 0] == chromosome]
        chromosome_starts[i, 0] = chromosome  # the chromosome
        chromosome_starts[i, 1] = chr_start_next  # start of the chromosome
        chromosome_starts[i, 2] = (
            chr_start_next + pos_chr[:, 2].max() ### pos_chr[:, 2]
        )  # end of the chromosome
        chr_start_next = chromosome_starts[i, 2] + offset
    return chromosome_starts
```

**Comment by vb-iman** [comment]
Hello!

No, this does not give the expected result. The x and y-axis positions of the points are not correct. However, if I omit the second change (pos_chr.max() --> pos_chr[:,2].max()), and only keep the first change ("object" --> "float") the plot looks as expected.

Is it possible to do "object" --> "int" instead, so that the chromosomes numbers are not floats in the plot?

**Comment by CarlKCarlK** [comment]
Thanks for trying this!

OK, I may have it. Sadly, I can't use `int` in `_compute_x_positions_chrom` because some datasets have base pairs numbers so big that they generate an error about overflowing a C 'long' (which I guess numpy is using for int).

Instead, please ...
Change the  main plot function, xticks by adding `.astype(int)`. This should display the int values as you requested.
```python
        plt.xticks(
            chromosome_starts[:, 1:3].mean(1), chromosome_starts[:, 0].astype(int)
        )
```

update with the 3 changes marked with `#...`
```python
def _compute_x_positions_chrom(positions, offset=1e5):
    chromosomes = np.unique(positions[:, 0])
    chromosomes.sort()
    chromosome_starts = np.zeros((chromosomes.shape[0], 3), dtype="float") # float
    chr_start_next = 0
    for i, chromosome in enumerate(chromosomes):
        pos_chr = positions[positions[:, 0] == chromosome, 1] # added ",1]"
        chromosome_starts[i, 0] = chromosome  # the chromosome
        chromosome_starts[i, 1] = chr_start_next  # start of the chromosome
        # end of the chromosome
        chromosome_starts[i, 2] = chr_start_next + np.nanmax(pos_chr) # make this nanmax to any missing values.
        chr_start_next = chromosome_starts[i, 2] + offset
    return chromosome_starts
```

**Comment by vb-iman** [comment]
These changes worked! Thank you

**Comment by CarlKCarlK** [comment]
Greetings! Thanks for your bug reports. I have a new beta version of FaST-LMM that should fix all the recent issues with NumPy compatibility, etc.

Please try:
` pip install fastlmm==0.6.8b1`

**Comment by CarlKCarlK** [comment]
Fixed with v 0.6.8
